src/mlpack/core/data/tokenizers/BERT_tokenizer_impl.hpp

```diff
@@ -0,0 +1,276 @@
+    while (pos != (size_t)-1 && isStripChar(ret[pos])) pos--;
+    return ret.substr(0, pos + 1);
+}
+
+static std::vector<std::wstring> split(const std::wstring& text) {
+    std::vector<std::wstring>  result;
+    boost::split(result, text, boost::is_any_of(stripChar));
+    return result;
+}
+
+static std::vector<std::wstring> whitespaceTokenize(const std::wstring& text) {
+    std::wstring rtext = strip(text);
+    if (rtext.empty()) return std::vector<std::wstring>();
+    return split(text);
+}
+
+void convertToUnicode(const std::string &s, std::wstring &ws) {
+
+    std::wstring wsTmp(s.begin(), s.end());
+
+    ws = wsTmp;
+}
+
+void convertFromUnicode(const std::wstring &ws, std::string &s) {
+
+    std::string sTmp(ws.begin(), ws.end());
+
+    s = sTmp;
+}
+
+static std::shared_ptr<Vocab> loadVocab(const std::string& vocabFile) {
+    std::shared_ptr<Vocab> vocab(new Vocab);
+    size_t index = 0;
+
+    // Loading contents of vocab file from file stream object to vector<string>.
+    std::fstream f;
+    f.open(vocabFile);
+
+    std::string tkn;
+    std::vector<std::string> vocabset;
+    while (std::getline(f, tkn))
+    {
+        vocabset.push_back(tkn);
+    }
+
+    f.close();
+
+    // Loading contents of vocab file from vector<string> a shared_ptr pointing towards wstring variables.
+    for (size_t i = 0; i < vocabset.size(); ++i) {
+        std::wstring token;
+        // Converting std::string datatype to std::wstring datatype.
+        convertToUnicode(vocabset[i], token);
+        token = strip(token);
+        (*vocab)[token] = index;
+        index++;
+    }
+    return vocab;
+}
+
+std::wstring str_replacer(std::wstring str, std::wstring old_substring, std::wstring new_substring)
+{
+  size_t index = 0;
+  while (true) {
+     /* Locate the substring to replace. */
+     index = str.find(old_substring, index);
+     if (index == std::string::npos) break;
+
+     /* Make the replacement. */
+     str.replace(index, old_substring.size(), new_substring);
+
+     /* Advance index forward so the next iteration doesn't pick it up as well. */
+     //index += old_substring.size();
+     index += old_substring.size();
+  }
+  return str;
+}
+
+std::wstring BasicTokenizer::cleanText(const std::wstring& text) const {
+    std::wstring output;
+    for (const wchar_t& cp : text)  {
+        if (isWhitespace(cp)) output += L" ";
+        else output += cp;
+    }
+    return output;
+}
+
+bool BasicTokenizer::isWhitespace(const wchar_t& ch) const {
+    if (ch== L' ' || ch== L'\t' || ch== L'\n' || ch== L'\r') return true;
+    return false;
+}
+
+bool BasicTokenizer::isPunctuation(const wchar_t& ch) const {
+    if (ch == '!' || ch == ',' || ch == ';' || ch == '.' || ch == '?' ||   
+       ch == '-' || ch == '\'' || ch == '\"' || ch == ':' || ch == '(' ||
+       ch == ')' || ch == '[' || ch == ']' || ch == '{' || ch == '}' ) return true;
+    return false;
+}
+
+BasicTokenizer::BasicTokenizer(bool doLowerCase=true) 
```

Comment:
Please don't specify the default argument in the implementation.  That's really hard for users to find.